test/1_test_fullpath.js

```diff
@@ -23,13 +23,16 @@ contract('Voucher tests', async (addresses) => {
     contractCashier,
     contractBosonRouter,
     contractFundLimitsOracle;
-  let tokenSupplyKey1, tokenSupplyKey2, tokenVoucherKey1, tokenVoucherKey2;
+  let tokenSupplyKey1,
+    tokenSupplyKey2,
+    tokenVoucherKey1,
+    tokenVoucherKey2,
+    promiseId1,
+    promiseId2;
 
-  before('setup contracts for tests', async () => {
+  beforeEach('setup contracts for tests', async () => {
```

Comment:
Do we need to redeploy the contracts before each test? Can we try to avoid that where possible, so we can speed up the tests a little bit?

Reply:
We want to be able to run each test case (`it`) using `it.only` in addition to running the test cases as part of the whole script. For this to work, each test case needs to be independent of any other test case. It can't rely on  or be influenced by state changes from previous test cases, so fresh state is needed for each test case. Without fresh state for each test case, we run into problems, especially with test cases that check nonces and correlation Ids. The assertions will pass or fail depending on whether you ran them with `it.only` or as part of the whole script. If the test case has fresh state, the assertions will always pass because the values will always be the same (unless a bug is introduced in the code).

Redeploying each time does take longer, but you shouldn't have to run the full script but once for each change. For example, if you add a new function, you might have a happy path test case and an error/negative test case. You should be able to add your first test case as an `it` and test it using `it.only`. Then you move on to your second test case and do the same. Only after you are satisfied that they both work do you have to run the whole script to run all the unit tests.